src/middlewared/middlewared/plugins/mdns.py

```diff
@@ -240,36 +363,36 @@ def start(self):
         if not mDNSDaemonMonitor.instance.mdnsd_running.wait(timeout=10):
             return
 
-        try:
-            hostname = socket.gethostname().split('.')[0]
-        except IndexError:
-            hostname = socket.gethostname()
-
-        mdns_advertise_services = [
-            mDNSServiceSSHThread,
-            mDNSServiceSFTPThread,
-            mDNSServiceHTTPThread,
-            mDNSServiceHTTPSThread,
-            mDNSServiceMiddlewareThread,
-            mDNSServiceMiddlewareSSLThread
-        ]
-
-        for service in mdns_advertise_services:
-            thread = service(middleware=self.middleware, hostname=hostname)
-            thread.setup()
-            thread_name = thread.service
-            self.threads[thread_name] = thread
-            thread.start()
+        service_info = self.middleware.call_sync('service.query')
+        hostname = self.middleware.call_sync('mdnsadvertise.get_hostname')
+        if hostname is None:
+            return
+
+        if self.threads.get('mDNSThread'):
+            self.logger.debug('mDNS advertise thread is already started.')
+            return
+
+        thread = mDNSServiceThread(
+            middleware=self.middleware,
+            hostname=hostname,
+            service_info=service_info
+        )
+        thread.setup()
+        thread_name = 'mDNSThread'
+        self.threads[thread_name] = thread
+        thread.start()
 
         with self.lock:
             self.initialized = True
 
     @private
     def stop(self):
-        for thread in self.threads.copy():
-            thread = self.threads.get(thread)
-            thread.cancel()
-            del self.threads[thread.service]
+        thread = self.threads.get('mDNSThread')
+        if thread is None:
+            return
+
+        thread.cancel()
+        del self.threads['mDNSThread']
```

Comment:
That seems unnecessary if we are setting threads below. The object will be garbage collected anyway.

Reply:
Removed.